Streaming lake ingestion private preview user's guide (#4004)
SoftwareAG/c8y-docs: content/datahub/streaming-lake-ingestion-bundle/using.md

---
weight: 20
title: Using Streaming Lake Ingestion
layout: redirect
---

> Streaming Lake Ingestion is an optional service in {{< product-c8y-iot >}}. To use the service, subscribe to it. After subscription, the service automatically stores all newly incoming data in the lake.

Reviewer: Should we add a description or link here on how to subscribe?

Author: AFAIK there is no self-service subscription yet. I'll change it to contact support.

{{< c8y-admon-info >}}

> On subscription, master data is synchronized with the operational store. This synchronization may take a while to complete. For more information, see "[Monitoring the data lake storage](#monitoring-the-data-lake-storage)".

Reviewer: ```suggestion
On subscription, master data is synchronized with the operational store. This synchronization may take a while to complete. For more information, see [Monitoring the data lake storage](#monitoring-the-data-lake-storage).
```

> The service stores only new data incoming after subscription. It does not automatically move data stored in the {{< product-c8y-iot >}} operational store before subscription. For more information, see "[Migrating to data lake storage](#migrating-to-data-lake-storage)".

Reviewer: ```suggestion
The service stores only new data incoming after subscription. It does not automatically move data stored in the {{< product-c8y-iot >}} operational store before subscription. For more information, see [Migrating to data lake storage](#migrating-to-data-lake-storage).
```

> *This feature is not part of the current private preview release.*

Reviewer: ```suggestion
*This feature is not part of the current Private Preview release.*
```

{{< /c8y-admon-info >}}

> ### Analyzing lake data from Cockpit

Reviewer: ```suggestion
### Analyzing lake data from Cockpit (#analyzing-lake-data-from-cockpit)
```

> Use {{< product-c8y-iot >}} Cockpit to access data stored in the {{< product-c8y-iot >}} data lake. Log in to the application, then navigate to the data explorer and select *data lake*. When you enable data lake, the application sources data from the data lake instead of the operational store. You can visualize data over longer time ranges than with the operational store.

Reviewer: ```suggestion
Use the {{< product-c8y-iot >}} Cockpit application to access data stored in the {{< product-c8y-iot >}} data lake. Log in to the application, then navigate to the data explorer and select **Data lake**. When you enable data lake, the application sources data from the data lake instead of the operational store. You can visualize data over longer time ranges than with the operational store.
```

{{< c8y-admon-info >}}
Data arrives in the lake with a small delay. If you need to see the latest data, switch to the operational store.

> *This feature is not part of the current private preview release.*

Reviewer: ```suggestion
*This feature is not part of the current Private Preview release.*
```

{{< /c8y-admon-info >}}

> ### Analyzing lake data using SQL

Reviewer: ```suggestion
### Analyzing lake data using SQL (#analyzing-lake-data-using-sql)
```

For more general analytics, use SQL queries. Run queries from several interfaces:

* Through the user interface of the [embedded query engine](/datahub/setting-up-datahub/#dremio-api-user).
* By connecting query tools such as BI tools and database browsers through [JDBC or ODBC](/datahub/setting-up-datahub/#dremio-api-user).
* By using the secure and high performance [querying APIs](https://cumulocity.com/api/datahub/) from applications.

> ### Understanding the data lake structure

Reviewer 2: We should also describe out namespace structure. 
For now we have namespace with tenant id but that may/will be subject to change

The structure of the data lake provided by Streaming Lake Ingestion mirrors the familiar [{{< product-c8y-iot >}} domain model](/concepts/domain-model/) to be intuitive for users. As shown in the screenshot below, the top level consists of folders representing the {{< product-c8y-iot >}} domain model classes: `inventory`, `alarm`, `event`, `measurement` and `operation`.

Within these folders, tables store the incoming IoT data. For example, the `alarm` table contains a complete, historical log of all alarm changes, while the `inventory` table captures all modifications made to your device master data over time.

> Device- or customer-specific data in the form of "[fragments](/concepts/domain-model/#fragments)" resides in separate, dedicated tables within the corresponding folders. For instance, if you send a measurement of type `c8y_EngineMetric`, it will be stored in a table named `c8y_EngineMetric` inside the measurements folder.

Reviewer: ```suggestion
Device- or customer-specific data in the form of [fragments](/concepts/domain-model/#fragments) resides in separate, dedicated tables within the corresponding folders. For instance, if you send a measurement of type `c8y_EngineMetric`, it will be stored in a table named `c8y_EngineMetric` inside the measurements folder.
```

> To simplify querying the current state of your assets, a consolidated representation of that state is available in "latest tables". The `inventory_latest` and `alarms_latest` tables show the most recent, up-to-date state of your inventory and alarms, respectively, without you needing to reconstruct it from the historical logs.

Reviewer 3: Not present in current preview.

Author: Yes, do you have a docs issue for updating the folder structure/naming? :-)

Reviewer 3: I was rather thinking of adding a note here that it's not available yet in the first place.

> To simplify querying the current state of your assets, a consolidated representation of that state is available in "latest tables". The `inventory_latest` and `alarms_latest` tables show the most recent, up-to-date state of your inventory and alarms, respectively, without you needing to reconstruct it from the historical logs.

Reviewer 2: current convention is that cdc and latest tables are placed in separate  namespaces

![alt text](/images/datahub-guide/querying.png)

To understand how the service transfers data to the data lake, the following sections discuss the general table structure for each {{< product-c8y-iot >}} domain model class.

> #### General table structure

Reviewer: ```suggestion
#### General table structure (#general-table-structure)
```

All tables contain the following key columns:
* `eventType`: The {{< product-c8y-iot >}} real-time data feed records changes to IoT data, so whether new data is created, existing data is updated or data is deleted. This field captures the type of the recorded change, for example, `MANAGED_OBJECT_UPDATE`.
* `id` resp. `source`: The [managed object ID](/concepts/domain-model/#object-identification), so the {{< product-c8y-iot >}} ID of the device or asset associated with the change.
* `time`: The timestamp of the change as sent by the device, if present, otherwise the [time when the change reaches {{< product-c8y-iot >}}](/device-management-application/monitoring-and-controlling-devices/#to-monitor-the-connection-of-a-particular-device).
* `type`: The type property of the device, measurement and so on.

> #### Inventory

Reviewer: ```suggestion
#### Inventory (#inventory)
```

* `type` contains "sb_nano", the type of the device as used in the request to create the device.
* `name` contains "Tracker #1", the name of the device as used in the request.
* `fragments` contains a list of all fragments used in the managed object.
* `someproperty` contains 10. {{< product-c8y-iot >}} creates additional columns in the `inventory` table when there are top-level properties of an atomic type in the managed object (character strings, numbers, booleans).

{{< c8y-admon-info >}}
Use top-level properties with care. For example, if you have custom properties that you want to store for all devices, such properties may be a good choice.
However, the total number of top-level properties is limited through the [table width limit of Iceberg](https://cumulocity.com/docs/service-terms/quotas/). If you exceed the table width limit,
additional properties will not be written to the inventory table but "binned" (see below).
Note also that very wide tables with many properties will lead to more inefficient storage and querying.
We advice you to prefer fragments for custom data when modeling your device data model.
{{< /c8y-admon-info >}}

**Table: inventory.c8y_position**

> | eventType             | id    | time                     | type    | alt | lng     | lat       |

Reviewer 2: ```suggestion
| eventType             | source    | time                     | type    | alt | lng     | lat       |
```
We use source also at fragments

| --------------------- | ----- | ------------------------ | ------- | --- | ------- | --------- |
| MANAGED_OBJECT_CREATE | 47635 | 2025-08-20T13:41:39.678Z | sb_nano | 67  | 6.15173 | 51.211977 |

> The columns of the `c8y_position` table reflect the "c8y_Position" fragment and are set as follows:

Reviewer: ```suggestion
The columns of the `c8y_position` table reflect the "`c8y_Position` fragment and are set as follows:
```

* Key columns are set as previously described.
* Properties are stored in the columns of the table.

In addition, the service stores the data in the `inventory_latest` folder:

**Table: inventory_latest.inventory**
| id    | time                     | type    | name       | fragments                      | someproperty |
| ----- | ------------------------ | ------- | ---------- | ------------------------------ | ------------ |
| 47635 | 2025-08-20T13:41:39.678Z | sb_nano | Tracker #1 | \[c8y_IsDevice, c8y_Position\] | 10           |

**Table: inventory_latest.c8y_position**

> | id    | time                     | type    | alt | lng     | lat       |

Reviewer 2: ```suggestion
| source    | time                     | type    | alt | lng     | lat       |
```

| ----- | ------------------------ | ------- | --- | ------- | --------- |
| 47635 | 2025-08-20T13:41:39.678Z | sb_nano | 67  | 6.15173 | 51.211977 |

{{< c8y-admon-info >}}

> "Fragments" such as "c8y\_IsDevice" are only represented in the "fragments" property of the `inventory` tables.

Reviewer: ```suggestion
Fragments such as `c8y\_IsDevice` are only represented in the "fragments" property of the `inventory` tables.
```

Reviewer: Shouldn´t it be `c8y_IsDevice` (without /)?

{{< /c8y-admon-info >}}

Assume that the tracking device updates its location:
```
{
    "c8y_Position": {
        "alt": 69,
        "lng": 6.3213,
        "lat": 50.425
    },
}
```

This update appears in the `inventory` and `c8y_position` tables as follows:

**Table: inventory.inventory**
| eventType             | id    | time                     | type    | name       | fragments                      | someproperty |
| --------------------- | ----- | ------------------------ | ------- | ---------- | ------------------------------ | ------------ |
| MANAGED_OBJECT_CREATE | 47635 | 2025-08-20T13:41:39.678Z | sb_nano | Tracker #1 | \[c8y_IsDevice, c8y_Position\] | 10           |
| MANAGED_OBJECT_UPDATE | 47635 | 2025-08-20T13:45:20.002Z | sb_nano | Tracker #1 | \[c8y_IsDevice, c8y_Position\] | 10           |

**Table: inventory.c8y_position**

> | eventType             | id    | time                     | type    | alt | lng     | lat       |

Reviewer 2: ```suggestion
| eventType             | source    | time                     | type    | alt | lng     | lat       |
```

| --------------------- | ----- | ------------------------ | ------- | --- | ------- | --------- |
| MANAGED_OBJECT_CREATE | 47635 | 2025-08-20T13:41:39.678Z | sb_nano | 67  | 6.15173 | 51.211977 |
| MANAGED_OBJECT_UPDATE | 47635 | 2025-08-20T13:45:20.002Z | sb_nano | 69  | 6.3213  | 50.425    |

The "latest" versions of the tables reflect this update:

**Table: inventory_latest.inventory**
| id    | time                     | type    | name       | fragments                      | someproperty |
| ----- | ------------------------ | ------- | ---------- | ------------------------------ | ------------ |
| 47635 | 2025-08-20T13:45:20.002Z | sb_nano | Tracker #1 | \[c8y_IsDevice, c8y_Position\] | 10           |

**Table: inventory_latest.c8y_position**

> | id    | time                     | type    | alt | lng    | lat    |

Reviewer 2: ```suggestion
| source    | time                     | type    | alt | lng    | lat    |
```

| ----- | ------------------------ | ------- | --- | ------ | ------ |
| 47635 | 2025-08-20T13:45:20.002Z | sb_nano | 69  | 6.3213 | 50.425 |

Finally, assume that you delete the device. This results in a record with `eventType` "MANAGED_OBJECT_DELETE" and the last properties stored on the device. Deleting the device removes it from the "latest" versions of the tables.

**Table: inventory.inventory**
| eventType             | id    | time                     | type    | name       | fragments                      | someproperty |
| --------------------- | ----- | ------------------------ | ------- | ---------- | ------------------------------ | ------------ |
| MANAGED_OBJECT_CREATE | 47635 | 2025-08-20T13:41:39.678Z | sb_nano | Tracker #1 | \[c8y_IsDevice, c8y_Position\] | 10           |
| MANAGED_OBJECT_UPDATE | 47635 | 2025-08-20T13:45:20.002Z | sb_nano | Tracker #1 | \[c8y_IsDevice, c8y_Position\] | 10           |
| MANAGED_OBJECT_DELETE | 47635 | 2025-08-20T13:50:20.002Z | sb_nano | Tracker #1 | \[c8y_IsDevice, c8y_Position\] | 10           |

**Table: inventory.c8y_position**

> | eventType             | id    | time                     | type    | alt | lng     | lat       |

Reviewer 2: ```suggestion
| eventType             | source    | time                     | type    | alt | lng     | lat       |
```

| --------------------- | ----- | ------------------------ | ------- | --- | ------- | --------- |
| MANAGED_OBJECT_CREATE | 47635 | 2025-08-20T13:41:39.678Z | sb_nano | 67  | 6.15173 | 51.211977 |
| MANAGED_OBJECT_UPDATE | 47635 | 2025-08-20T13:45:20.002Z | sb_nano | 69  | 6.3213  | 50.425    |
| MANAGED_OBJECT_DELETE | 47635 | 2025-08-20T13:50:20.002Z | sb_nano | 69  | 6.3213  | 50.425    |

> #### Alarms

Reviewer: ```suggestion
#### Alarms (#alarms)
```

| ALARM_CREATE | 47635  | 2025-08-19T12:03:27.845Z | c8y_BatteryAlarm | 1     | MAJOR    | ACTIVE | Battery level below 5 percent. |

Again, the columns of the table represent the [properties of a {{< product-c8y-iot >}} alarm](https://cumulocity.com/api/core/#tag/Alarms). The `eventType` property reflects the recorded change (in this case, a created alarm), and the `count` property is added by {{< product-c8y-iot >}}'s [alarm de-duplication](https://cumulocity.com/api/core/#alarm-de-duplication).

In the example, there are no custom properties or fragments in the alarm data. The service treats such properties or fragments in the same way as described for the inventory.

For alarms, there is also a representation of the latest state.

**Table: alarm_latest.alarm**
| source | time                     | type             | count | severity | status | text                           |
| ------ | ------------------------ | ---------------- | ----- | -------- | ------ | ------------------------------ |
| 47635  | 2025-08-19T12:03:27.845Z | c8y_BatteryAlarm | 1     | MAJOR    | ACTIVE | Battery level below 5 percent. |

> #### Events

Reviewer: ```suggestion
#### Events (#events)
```

```
{
    "c8y_Position": { "alt": 67, "lng": 6.95173, "lat": 51.151977 },
    "time": "2025-08-20T13:41:39.678Z",
    "source": { "id": "47635" },
    "type": "c8y_LocationUpdate",
    "text": "Tracker location update"
}
```

This results in the following data in the data lake, consistent with the method described previously.

**Table: event.event**
| eventType    | source | time                     | type               | text                    |
| ------------ | ------ | ------------------------ | ------------------ | ----------------------- |
| EVENT_CREATE | 47635  | 2025-08-20T13:41:39.678Z | c8y_LocationUpdate | Tracker location update |

**Table: event.c8y_position**
| eventType    | id    | time                     | type               | alt | lng     | lat       |
| ------------ | ----- | ------------------------ | ------------------ | --- | ------- | --------- |
| EVENT_CREATE | 47635 | 2025-08-20T13:41:39.678Z | c8y_LocationUpdate | 67  | 6.15173 | 51.211977 |

As events usually record specific occurrences or state changes for a device at a single point in time and are distinct, the service does not create a "latest" version.

> #### Measurements

Reviewer: ```suggestion
#### Measurements (#measurements)
```

```
{
  "time": "2025-08-20T13:42:39.678Z",
  "source": { "id": "47635" },
  "type": "c8y_BatteryMeasurement",
  "c8y_Battery": {
    "voltage": { "value": 12.8, "unit": "V" },
    "stateOfCharge": { "value": 85.5, "unit": "%" },
    "temperature": { "value": 22.5, "unit": "C" }
  }
}
```

This results in the following data in the data lake:

**Table: measurement.c8y_Battery**
| eventType          | source | time                     | type                   | voltage       | stateOfCharge | temperature   |
| ------------------ | ------ | ------------------------ | ---------------------- | ------------- | ------------- | ------------- |
|                    |        |                          |                        | value \| unit | value \| unit | value \| unit |
| MEASUREMENT_CREATE | 47635  | 2025-08-20T13:42:39.678Z | c8y_BatteryMeasurement | 12.8 \| V     | 85.5 \| %     | 22.5 \| C     |

There is one important difference for measurements compared to other data types: There is no `measurement.measurement` table. All relevant data is present in the fragment tables.

> This example demonstrates how the service handles nested properties: Like the {{< product-c8y-iot >}} APIs and its operational store, Apache Iceberg supports nested data. Just like with the above JSON example, the property `voltage` is a nested structure containing two properties `value` and `unit`. You can query the contained properties from SQL using path expressions such as `"voltage"."value"`. For more information on querying the data from SQL, see the Section "[Example queries](#example-queries)".

Reviewer: ```suggestion
This example demonstrates how the service handles nested properties: Like the {{< product-c8y-iot >}} APIs and its operational store, Apache Iceberg supports nested data. Just like with the above JSON example, the property `voltage` is a nested structure containing two properties `value` and `unit`. You can query the contained properties from SQL using path expressions such as `"voltage"."value"`. For more information on querying the data from SQL, see [Example queries](#example-queries).
```

> #### Operations

Reviewer: ```suggestion
#### Operations (#operations)
```

The columns represent the [properties of a {{< product-c8y-iot >}} operation](https://cumulocity.com/api/core/#operation/getOperationCollectionResource). If you do not provide an operational property, the service stores it as a SQL "null" value.

#### Data types

{{< product-c8y-iot >}} relies on JSON as data exchange format in its APIs. JSON data types map to Iceberg data types as follows:

* Strings parsed as timestamps become the Iceberg `timestamptz` data type. Otherwise, they become the Iceberg `string` data type.
* Numbers are stored as Iceberg `decimal` with a maximum precision of 38 and scale of 9.
* Booleans (true/false) map to the equivalent Iceberg `boolean` data type.
* JSON lists become the equivalent Iceberg `list` data type.
* JSON objects become the equivalent Iceberg `struct` data type.

"Null" values are stored as Iceberg "null" values.

{{< c8y-admon-info >}}
For more information related to Iceberg data types, please refer to the [Iceberg specification](https://iceberg.apache.org/spec/#semi-structured-types). Note that there are various [structural limits in Iceberg](/service-terms/quotas/).
{{< /c8y-admon-info >}}

> #### Schema evolution

Reviewer: ```suggestion
#### Schema evolution (#schema-evolution)
```

The data lake storage service automatically discovers the structure of the incoming data and applies it to the data lake.
* When the service discovers a new fragment, it creates a table to store the data in the fragment.
* When the service discovers a new property, it creates a column to store the data in the associated table.

{{< c8y-admon-info >}}

> Currently, the service does not delete tables and columns when they are not in use anymore. For more information, see Section "[Managing schema evolution](#managing-schema-evolution)".

Reviewer: ```suggestion
Currently, the service does not delete tables and columns when they are not in use anymore. For more information, see [Managing schema evolution](#managing-schema-evolution).
```

{{< /c8y-admon-info >}}

> #### Handling of schema conflicts

Reviewer: ```suggestion
#### Handling of schema conflicts (#handling-of-schema-conflicts)
```

| decimal               | __d    |
| timestamp             | __t    |
| boolean               | __b    |
| struct                | __o    |
| list&lt;string&gt;    | __ls   |
| list&lt;decimal&gt;   | __ld   |
| list&lt;timestamp&gt; | __lt   |
| list&lt;boolean&gt;   | __lb   |
| list&lt;struct&gt;    | __lo   |

If you send additional type variations for the same property, additional overflow columns are created accordingly. For example, if "Tracker #3" later sends `"someProperty": [true]`, a new column `someproperty__lb` would be created for the list of booleans.

{{< c8y-admon-info >}}
The type of the existing column never changes automatically to ensure that your existing queries and reports do not break. Also, to prevent naming conflicts, incoming data with property names ending in these overflow suffixes is rejected and not stored in the main tables.

If there is a conflict between a top-level property with a fragment of the same name, the service creates a new table for the fragment. For example, if you have a top-level `"version": "1.0"` and send a `"version": { "major": 1, "minor": 0 }`, a table `version` is created.
{{< /c8y-admon-info >}}

{{< c8y-admon-caution >}}

> The service currently does not support certain combinations of type conflicts with lists. It moves conflicting lists to the trash table.

Reviewer 3: This is the first mention of "trash table". We only use "binning" in other places.

Reviewer 4: `trash` table is elaborated under "Binning" section already. We can highlight `trash` word here and comment as see below/link to "Binning" section.

{{< /c8y-admon-caution >}}

> #### Naming

Reviewer: ```suggestion
#### Naming (#naming)
```

the [Apache Avro specification](https://avro.apache.org/docs/1.8.1/spec.html#names) and Iceberg's Avro support.
{{< /c8y-admon-info >}}

Characters are case-sensitive, but identifiers of tables and columns must be unique in a case-insensitive way.
When the column or table name is not unique, a unique suffix is automatically appended to the name.

The suffix is generated based on the original name as a binary code where uppercase letters are represented as 1 and lowercase letters as 0. This binary code is then converted to a hexadecimal number and appended to the name in the format `{name}__{suffix}`. For example:

| Input  | Binary | Hex | Output     |
| ------ | ------ | --- | ---------- |
| userId | 000010 | 2   | userId__2  |
| UserId | 100010 | 22  | UserId__22 |
| USERID | 111111 | 3F  | UserId__3F |
| userid | 000000 | 0   | userId__0  |
| uSeRiD | 010101 | 15  | userId__15 |

{{< c8y-admon-info >}}
This is a reliability measure. The Iceberg specification considers tables and column names as case sensitive. However, upstream query engines do not generally support case sensitivity and may not be able to query such tables.
{{< /c8y-admon-info >}}

> #### Flat views

Reviewer: ```suggestion
#### Flat views (#flat-views)
```

{{< c8y-admon-info >}}
For legacy analytics tools, {{< company-c8y >}} plans to offer Iceberg views with a flat representation of the Iceberg tables. This is currently under implementation.
{{< /c8y-admon-info >}}

> #### Binning

Reviewer: ```suggestion
#### Binning (#binning)
```

| ------ | --------------------- | --------- | -------------------- | ----------------------------------------------------------- | -------------------- | ------------------ | ----------------------------------------- | ---------------------------------- |
| 47635  | EVENT_CREATE          | EVENT     | 2025-07-29T13:15:43Z | /event_fragment_1/c8y_TooLargeProperty                      | event_fragment_1     | SIZE_EXCEEDED      | Property size exceeded the limit of 32768 | IuS4gOefs+S6jOmzpeS4gOefs+S....... |
| 47638  | MANAGED_OBJECT_CREATE | INVENTORY | 2025-07-29T14:15:43Z | /inventory_fragment_2/level_1/level_2....level_17/name      | inventory_fragment_2 | MAX_DEPTH_EXCEEDED | Property depth exceeded the limit of 16   | IlRoZSBhYnlzcyI=                   |
| 47640  | EVENT_CREATE          | EVENT     | 2025-07-29T15:15:43Z | /event_fragment_2/nestedObject/invalidFieldName###ASD!ASD!@ | event_fragment_2     | ILLEGAL_FIELD_NAME | Illegal field name                        | InRlc3Qi                           |

> #### Customer-visible operational communication {#customer-visible-operational-communication}

Author: @rahj-c8y Please move the section to "Monitoring the data flow" and phrase it as customer documentation rather than as technical specification. 

Please have a quick look at the docs style guide or the copilot-instructions in this repo. Customers are directly addressed with "you" instead of "the customer". 

Please list concretely the events and alarms that you raise, describing the circumstances under which they appear, preferably with one screenshot from the event viewer and alarm UI rather than the JSON. We might add the JSON lateron (but let's review the concrete structure of it first and check if it complies with our general guidelines -- see other PR comment).

---
weight: 20
title: Using Streaming Lake Ingestion
layout: redirect
---

Streaming Lake Ingestion is an optional service in {{< product-c8y-iot >}}. To subscribe to the service, contact the [{{< company-c8y >}} support](/additional-resources/contacting-support/). After subscription,

* Your current device and asset inventory is downloaded from the operational store into the lake.
* All new incoming data is stored in the lake.

{{< c8y-admon-info >}}
The download may take a while to complete. For more information, see [Monitoring the data lake storage](#monitoring-the-data-lake-storage).

> The `latest_inventory` tables are pre-populated with your full current inventory at subscription time. Alarms, events, measurements and operations are only recorded for changes that occur after subscription.

Reviewer: ```suggestion
The `latest_inventory` tables are pre-populated with your full current inventory at subscription time. Alarms, events, measurements, and operations are only recorded for changes that occur after subscription.
```

For users of DataHub Query (Dremio), the option `CDH_ASSET_WHITELIST` is automatically set on your tenant. This allows granting the tenant's default Dremio user access to the Iceberg source, which is required to read data via the DataHub REST API. Do not change this option.
{{< /c8y-admon-info >}}

### Analyzing lake data {#analyzing-lake-data-using-sql}

You can now analyze your data from any Apache Iceberg compliant tool. For example:

* Through the user interface of the [embedded query engine](/datahub/setting-up-datahub/#dremio-api-user).
* By using the secure and high performance [querying APIs](https://cumulocity.com/api/datahub/) from applications and by using JDBC or ODBC with query engines and database tools.

> * By connecting applications directly to the Cumulocity Iceberg catalog.

Reviewer: ```suggestion
* By connecting applications directly to the {{< product-c8y-iot >}} Iceberg catalog.
```

> #### Using DataHub Query to analyze lake data

Reviewer: This and the following heading is missing an ID.

For example, use the REST API to submit a query job to DataHub Query.

```shell
$ curl -H "Content-Type: application/json" -u "<USER>:<PASS>" \
  https://<TENANT_DOMAIN>/service/datahub/dremio/api/v3/sql -d @- <<EOF \
{
  "sql": "SELECT * FROM inventory",
  "context": [ "<TENANT>", "cdc_inventory" ]
}
EOF

{"id":"12345678-1234-5678-1234-5678901234567"}
```

Then fetch the results of the query job once the job has finished.

```shell
$ curl -u "admin:$PASS" \
  https://<TENANT_DOMAIN>/service/datahub/dremio/api/v3/job/12345678-1234-5678-1234-5678901234567/results

{
  "rowCount": 7,
  "schema": [
    { "name": "id", "type": { "name": "VARCHAR" } },
    …
  ],
  "rows": [
    {
      "id": "99266201",
      "lastUpdated": "2026-03-30 15:27:35.781",
      "name": "Temperature #1",
      "owner": "service_device-simulator",
      "type": "c8y_MQTTDevice",
      "eventType": "MANAGED_OBJECT_UPDATE",
    …
  ]
}
```

#### Obtaining Iceberg catalog credentials {#obtaining-iceberg-catalog-credentials}

> To connect to the Cumulocity Iceberg catalog directly — for example, from Apache Spark, Databricks, or a custom application — you need OAuth2 client credentials. As a tenant administrator, you can create and manage named catalog principals using the Manager API.

Reviewer: ```suggestion
To connect to the {{< product-c8y-iot >}} Iceberg catalog directly — for example, from Apache Spark, Databricks, or a custom application — you need OAuth2 client credentials. As a tenant administrator, you can create and manage named catalog principals using the Manager API.
```

**Prerequisites**

> * Your Cumulocity user must have the `ROLE_TENANT_ADMIN` role.

Reviewer: ```suggestion
* Your {{< product-c8y-iot >}} user must have the ROLE_TENANT_ADMIN permission.
```

  "clientSecret": "<SECRET>"
}
```

{{< c8y-admon-important >}}
The `clientSecret` is returned only once and is never stored by the service. Store it securely immediately after creation. If the secret is lost, rotate the principal to issue new credentials using a `PUT` request — rotating immediately invalidates the previous secret.
{{< /c8y-admon-important >}}

**Listing principals**

To list the names of all principals for your tenant:

```shell
curl -s \
  "https://<TENANT_DOMAIN>/service/offloading/api/v1/principals" \
  -u "<USER>:<PASS>"
```

**Rotating credentials**

To replace a lost or compromised secret. The previous secret is immediately invalidated.

```shell
curl -s -X PUT \
  "https://<TENANT_DOMAIN>/service/offloading/api/v1/principals/<NAME>" \
  -u "<USER>:<PASS>"
```

**Deleting a principal**

To revoke a principal's catalog access immediately:

```shell
curl -s -X DELETE \
  "https://<TENANT_DOMAIN>/service/offloading/api/v1/principals/<NAME>" \
  -u "<USER>:<PASS>"
```

{{< c8y-admon-info >}}
A maximum of 100 principals can be created per tenant by default.
{{< /c8y-admon-info >}}

#### Using the Iceberg catalog from Apache Spark

> [Apache Spark](https://spark.apache.org/) is a distributed computing framework that seamlessly integrates with the Cumulocity Iceberg catalogs to provide full SQL-based data processing through a standard [Iceberg REST catalog interface](https://iceberg.apache.org/rest-catalog-spec/).

Reviewer: ```suggestion
[Apache Spark](https://spark.apache.org/) is a distributed computing framework that seamlessly integrates with the {{< product-c8y-iot >}} Iceberg catalogs to provide full SQL-based data processing through a standard [Iceberg REST catalog interface](https://iceberg.apache.org/rest-catalog-spec/).
```

> Use [OpenID Connect](https://openid.net/developers/how-connect-works/) with a client credentials grant type to authenticate against the Cumulocity Iceberg REST catalog.

Reviewer: ```suggestion
Use [OpenID Connect](https://openid.net/developers/how-connect-works/) with a client credentials grant type to authenticate against the {{< product-c8y-iot >}} Iceberg REST catalog.
```

{{< c8y-admon-info >}}
To obtain client credentials for the Iceberg catalog, see [Obtaining Iceberg catalog credentials](#obtaining-iceberg-catalog-credentials).
{{< /c8y-admon-info >}}

The URL of the OpenID Connect server is

`https://iceberg.<INSTANCE>:19120/api/catalog/v1/oauth/tokens`

For example, if your tenant is `mytenant.cumulocity.com`, the URL is

`https://iceberg.cumulocity.com:19120/api/catalog/v1/oauth/tokens`.

> Run Spark SQL queries with the Cumulocity Iceberg catalog on an AWS S3 object store using the example below. In the example, replace

Reviewer: ```suggestion
Run Spark SQL queries with the {{< product-c8y-iot >}} Iceberg catalog on an AWS S3 object store using the example below. In the example, replace
```

  "endpoints": [
    "GET /v1/{prefix}/namespaces",
    "GET /v1/{prefix}/namespaces/{namespace}",
    "HEAD /v1/{prefix}/namespaces/{namespace}",
    "POST /v1/{prefix}/namespaces",
    "POST /v1/{prefix}/namespaces/{namespace}/properties",
    "DELETE /v1/{prefix}/namespaces/{namespace}",
    "GET /v1/{prefix}/namespaces/{namespace}/tables",
    "GET /v1/{prefix}/namespaces/{namespace}/tables/{table}",
    …
}

$ curl https://iceberg.<INSTANCE>:19120/api/catalog/v1/<TENANT>/namespaces/cdc_inventory/tables/inventory' \
  -H "Authorization: Bearer <ACCESS_TOKEN>"

{
  "metadata": {
    "format-version": 2,
    …
    "schemas": [
      {
        "type": "struct",
        "schema-id": 0,
        "fields": [
          { "id": 1, "name": "id", "required": true, "type": "string" },
          { "id": 2, "name": "lastUpdated", "required": true, "type": "timestamptz" },
          { "id": 3, "name": "name", "required": false, "type": "string" },
    …
```

{{< c8y-admon-info >}}
While Iceberg is widely supported, the degree of support currently still varies and we cannot guarantee the catalog to be interoperable with all setups and applications.
{{< /c8y-admon-info >}}

> ### Understanding the data lake structure

Reviewer: Some more headings are missing IDs, please check throughout this document.

Streaming Lake Ingestion provides three types of data:

* Change data capture (`cdc`): Includes the complete history of incoming IoT data. For example, the `cdc_alarm.alarm` table contains a complete, historical log of all alarm changes, while the `cdc_inventory.inventory` table captures all modifications made to your device master data over time.
* Latest data (`latest`): Provides the current representation of incoming inventory data. For example, the `latest_inventory.inventory` table contains a consolidated view on the latest state of your inventory without you needing to reconstruct the state from the historical logs.
* Views (`view`): Provides views for the change data capture tables. For example, the `view_alarm.alarm` view shows the same content as `cdc_alarm.alarm` with nested properties converted to top-level columns. This makes the data more easily browsable in tools like PowerBI.

> For each type of data, folders ("namespaces") are created to mirror the familiar [{{< product-c8y-iot >}} domain model](/concepts/domain-model/) of Cumulocity as shown in the screenshot. You will find folders with names ending in `inventory`, `alarm`, `event`, `measurement` and `operation`. Inside the folders, tables for standard data, device-specific data and customer-specific data are created. Device- or customer-specific data in the form of [fragments](/concepts/domain-model/#fragments) resides in separate, dedicated tables within the corresponding folders. For instance, if you send a measurement with a fragment type `c8y_EngineMetric`, it will be stored in a table named `cdc_measurement.c8y_EngineMetric`.

Reviewer: ```suggestion
For each type of data, folders ("namespaces") are created to mirror the familiar [{{< product-c8y-iot >}} domain model](/concepts/domain-model/) as shown in the screenshot. You will find folders with names ending in `inventory`, `alarm`, `event`, `measurement` and `operation`. Inside the folders, tables for standard data, device-specific data, and customer-specific data are created. Device- or customer-specific data in the form of [fragments](/concepts/domain-model/#fragments) resides in separate, dedicated tables within the corresponding folders. For instance, if you send a measurement with a fragment type `c8y_EngineMetric`, it will be stored in a table named `cdc_measurement.c8y_EngineMetric`.
```

The following sections discuss the general table structure for each {{< product-c8y-iot >}} domain model class.

#### General table structure {#general-table-structure}

All tables contain the following column:
* `id`: The record ID of the device, asset, alarm, event, measurement or operation associated with the change.

>   In the inventory tables, this is the [managed object ID](https://cumulocity.com/docs/concepts/domain-model/#object-identification).

Reviewer: ```suggestion
  In the inventory tables, this is the [managed object ID](/concepts/domain-model/#object-identification).
```

The following columns are present in some tables, depending on the type of data:

> * `source`: For alarms, events or measurement, this columns contains the [managed object ID](https://cumulocity.com/docs/concepts/domain-model/#object-identification) of the device or asset associated with the change.

Reviewer: ```suggestion
* `source`: For alarms, events, or measurement, this column contains the [managed object ID](/concepts/domain-model/#object-identification) of the device or asset associated with the change.
```

> * `deviceId`: For operations, this columns contains the [managed object ID](https://cumulocity.com/docs/concepts/domain-model/#object-identification) of the device targeted for the operation.

Reviewer: ```suggestion
* `deviceId`: For operations, this column contains the [managed object ID](/concepts/domain-model/#object-identification) of the device targeted for the operation.
```

> * `time`: The timestamp of the change as sent by the device. This column is present for alarms, events and measurements.

Reviewer: ```suggestion
* `time`: The timestamp of the change as sent by the device. This column is present for alarms, events, and measurements.
```

| 47635                                                       | 2025-08-20T13:41:39.678Z                                             | Tracker #1                                                    | device_123                                                     | sb_nano                                                       | MANAGED_OBJECT_CREATE                                              | \[c8y_IsDevice, c8y_Position\]                                      | \[\]           | \[\]        | \[\]         | 2025-08-20T13:41:39.678Z                                               | 10           |

The columns of the `inventory` table are:
* `id` contains "47635", the managed object ID assigned to the new device by {{< product-c8y-iot >}}.
* `lastUpdated` contains "2025-08-20T13:41:39.678Z", the timestamp of the most recent update to the managed object in {{< product-c8y-iot >}}.
* `name` contains "Tracker #1", the name of the managed object as used in the request.

> * `owner` contains the user who created the managed object

Reviewer: ```suggestion
* `owner` contains the user who created the managed object.
```

* `type` contains "sb_nano", the type of the managed object as used in the request to create the managed object.
* `eventType` contains "MANAGED_OBJECT_CREATE", the event type for creating a new managed object.
* `fragments` contains a list of all fragments used in the managed object.
* `childAdditions` contains a list of IDs of services linked to a device.
* `childAssets` contains a list of IDs of child assets of an asset.
* `childDevices` contains a list of IDs of child devices of an asset or device.
* `someproperty` contains 10. {{< product-c8y-iot >}} creates additional columns in the `inventory` table when there are top-level properties of an atomic type in the managed object (character strings, numbers, booleans).

> Additional columns may be visible depending on your use of Cumulocity. For example, once the device has emitted measurements, you will see a column `supportedMeasurements` with a list of the measurement fragments sent by the device.

Reviewer: ```suggestion
Additional columns may be visible depending on your use of {{< product-c8y-iot >}}. For example, once the device has emitted measurements, you will see a column `supportedMeasurements` with a list of the measurement fragments sent by the device.
```

The columns represent the [properties of a {{< product-c8y-iot >}} operation](https://cumulocity.com/api/core/#operation/getOperationCollectionResource). If you do not provide an optional property, the service stores it as a SQL "null" value.

#### Data types

{{< product-c8y-iot >}} relies on JSON as data exchange format in its APIs. JSON data types map to Iceberg data types as follows:

* Strings parsed as timestamps become the Iceberg `timestamptz` data type. Otherwise, they become the Iceberg `string` data type.
* Numbers are stored as Iceberg `decimal` with a maximum precision of 38 and scale of 9.
* Booleans (true/false) map to the equivalent Iceberg `boolean` data type.
* JSON lists become the equivalent Iceberg `list` data type.
* JSON objects become the equivalent Iceberg `struct` data type.

"Null" values are stored as Iceberg "null" values.

Empty JSON objects (`{}`) and empty arrays (`[]`) carry no data to store. They are omitted from the input and treated the same as an absent or "null" property, rather than being converted to a value of the column's data type.

{{< c8y-admon-info >}}

> For more information related to Iceberg data types, please refer to the [Iceberg specification](https://iceberg.apache.org/spec/#semi-structured-types). Note that there are various [structural limits](#limits-of-streaming-lake-ingestion) that the service implements to ensure that common upstream query engines can interact with the data produced by the service.

Reviewer: ```suggestion
For more information related to Iceberg data types, refer to the [Iceberg specification](https://iceberg.apache.org/spec/#semi-structured-types). Note that there are various [structural limits](#limits-of-streaming-lake-ingestion) that the service implements to ensure that common upstream query engines can interact with the data produced by the service.
```

```json
{
    "fragment": {
        "listField": ["Some string", 5.1, false, null]
    }
}
```

### Example queries {#example-queries}

The following example queries demonstrate how to extract common metrics and insights from the data lake using the Dremio SQL dialect. To keep the queries concise, they assume that you set the query context to your tenant as described in [Analyzing lake data](#analyzing-lake-data-using-sql). This allows you to omit the tenant prefix in the `FROM` clauses.

#### Basic inventory queries

To list all devices registered in your tenant, use the `latest_inventory` tables to get the most recent state and filter for the marker fragment `c8y_IsDevice` in the `fragments` array.

```sql
SELECT id, name
FROM latest_inventory.inventory
WHERE ARRAY_CONTAINS(fragments, 'c8y_IsDevice')
```

To find the current location of all devices within a specific geographic boundary, query the `c8y_Position` fragment in the `latest_inventory` folder.

```sql
SELECT id, lat, lng
FROM latest_inventory.c8y_Position
WHERE lat BETWEEN 50.0 AND 52.0
  AND lng BETWEEN 6.0 AND 8.0
```

#### Basic time series queries

> To retrieve all events recorded for a specific device within a given timespan, query the `cdc_event.event` table. Note that `time` is a reserved word in Dremio SQL, so it needs to be quoted.

Reviewer: ```suggestion
To retrieve all events recorded for a specific device within a given timespan, query the `cdc_event.event` table. Note that `time` is a reserved word in Dremio SQL, so it must be quoted.
```
